api/src/api/users.rs

```diff
@@ -0,0 +1,116 @@
+use crate::models::response;
+use actix_web::{HttpResponse, delete, get, post, web};
+use futures::stream::StreamExt;
+use mongodb::{Collection, Database, bson::doc};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+
+#[derive(Deserialize, Clone)]
+pub struct UserPost {
+    pub username: String,
+    pub email: String,
+    pub password: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct UserInsert {
+    pub id: String,
+    pub username: String,
+    pub email: String,
+    pub password: String, // This will be hashed before insertion
+}
+
+#[post("/users")]
+pub async fn post_user(db: web::Data<Database>, body: web::Json<UserPost>) -> HttpResponse {
+    let user_collection: Collection<UserInsert> = db.collection("users");
+
+    // Create a new user document
+    // First, hash password
+    // TODO: Permissions?
+    let user_id = uuid::Uuid::new_v4().to_string();
+    let hashed_password =
+        bcrypt::hash(&body.password, bcrypt::DEFAULT_COST).expect("failed to hash password");
+    let user_insert = UserInsert {
+        id: user_id.clone(),
+        username: body.username.clone(),
+        email: body.email.clone(),
+        password: hashed_password,
+    };
+
+    // Save new user to database
+    match user_collection.insert_one(user_insert).await {
+        Ok(_) => response::ok(
+            "success",
+            serde_json::to_value(User {
+                id: user_id,
+                username: body.username.clone(),
+                email: body.email.clone(),
+            })
+            .unwrap(),
+        ),
+        // Catch unique index constraint error
+        Err(e) if e.to_string().contains("E11000 duplicate key error") => HttpResponse::Conflict()
+            .body(format!(
+                "user with username '{}' already exists",
```

Comment:
This could technically be incorrect if we had a UUID collision, but I don't think that's going to happen with the number of users we expect here.